File: jianzhioffer/firstround/jianzhi_55_2.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


# class Solution(object):
#     def dfs(self, node):
#         if not node:
#             return 0
#         return max(self.dfs(node.left), self.dfs(node.right)) + 1
#
#     def isBalanced(self, root):
#         """
#         :type root: TreeNode
#         :rtype: bool
#         """
#         if not root:
#             return True
#         return abs(self.dfs(root.left) - self.dfs(root.right)) <= 1 and self.isBalanced(root.left) \
#                and self.isBalanced(root.right)


class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def recur(node):
            if not node:
                return 0
            left = recur(node.left)
            if left == -1:
                return -1
            right = recur(node.right)
            if right == -1:
                return -1
            return max(left, right) + 1 if abs(left - right) <= 1 else -1

        return recur(root) != -1


if __name__ == '__main__':
    n1 = TreeNode(3)
    n2 = TreeNode(9)
    n3 = TreeNode(20)
    n4 = TreeNode(15)
    n5 = TreeNode(7)
    # n6 = TreeNode(1)

    n1.left = n2
    n1.right = n3
    n3.left = n4
    n3.right = n5

    obj = Solution()
    isBalance = obj.isBalanced(n1)
    print(isBalance)

File: jianzhioffer/secondround/jianzhi_55_2.cpp
#include <cstddef>
#include <cstdlib>
#include <algorithm>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x):val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
    bool isBalanced(TreeNode* root);

private:
    int DFS(TreeNode* root);
};

//DFS，此处用后序遍历。时间复杂度O(N)，空间复杂度O(N)
bool Solution::isBalanced(TreeNode* root)
{
    if(root==NULL) return true;
    return DFS(root)!=-1;
}

int Solution::DFS(TreeNode* root)
{
    if(root==NULL) return 0;
    int left_depth = DFS(root->left);
    if(left_depth==-1) return -1;

    int right_depth = DFS(root->right);
    if(right_depth==-1) return -1;

    if(abs(left_depth-right_depth)>1) return -1;
    else return max(left_depth, right_depth)+1;
}

int main(int argc, char const *argv[])
{
    TreeNode* n1 = new TreeNode(3);
    TreeNode* n2 = new TreeNode(9);
    TreeNode* n3 = new TreeNode(20);
    TreeNode* n4 = new TreeNode(15);
    TreeNode* n5 = new TreeNode(7);

    n1->left = n2;
    n1->right = n3;
    n3->left = n4;
    n3->right = n5;
    Solution obj;
    int result = obj.isBalanced(n1);
    return 0;
}